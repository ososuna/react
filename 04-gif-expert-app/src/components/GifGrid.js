
import React, { useState, useEffect } from 'react'
import { getGifs } from '../helpers/getGifs';
import { GifGridItem } from './GifGridItem';

export const GifGrid = ({ category }) => {

    const [images, setImages] = useState([]);
    
    useEffect( async() =>{
        const images = await getGifs( category );
        setImages( images );
    }, [ category ]);
    
    return (
        <>
            <h3>{ category }</h3>
            <div className="card-grid">
                    {
                        images.map( ( img ) => (
                            <GifGridItem
                                key={ img.id }
                                { ...img }
                            />
                        ))
                    }
            </div>
        </>
    )
}
