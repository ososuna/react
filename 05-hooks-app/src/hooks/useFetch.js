
import { useEffect, useState } from "react"

export const useFetch = ( url ) => {

    if ( !url ) {
        throw new Error('The url is required');
    }

    const [state, setState] = useState({ data: null, loading: true, error: null });
    
    useEffect(() => {
        fetch( url )
            .then( resp => resp.json() )
            .then( data => {
                setState({
                    loading: false,
                    error: null,
                    data
                })
            });
    }, [ url ]);

    return state;

}
