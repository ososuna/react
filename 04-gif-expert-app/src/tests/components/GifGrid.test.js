
import React from 'react';
import { shallow } from "enzyme";
import '@testing-library/jest-dom';

import { GifGrid } from '../../components/GifGrid';
import { useFetchGifs } from '../../hooks/useFetchGifs';
jest.mock('../../hooks/useFetchGifs');

describe('GifGrid component tests', () => {
    
    const category = 'Hunter x Hunter';

    test('should match with snapshot', () => {

        useFetchGifs.mockReturnValue({
            data: [],
            loading: true
        });

        const wrapper = shallow( <GifGrid category={ category } /> );
        expect( wrapper ).toMatchSnapshot();
    
    });
    
    test('should show items when the images are loaded - useFetchGifs', () => {

        const gifs = [{
            id: '123',
            url: 'https://localhost/something.jpg',
            title: 'A title'
        }];

        useFetchGifs.mockReturnValue({
            data: gifs,
            loading: false
        });

        const wrapper = shallow( <GifGrid category={ category } /> );
        
        expect( wrapper ).toMatchSnapshot();
        expect( wrapper.find('p').exists() ).toBe( false );
        expect( wrapper.find('GifGridItem').length ).toBe( gifs.length );

    });
    
});
