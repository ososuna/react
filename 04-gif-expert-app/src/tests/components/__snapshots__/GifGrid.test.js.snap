// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GifGrid component tests should match with snapshot 1`] = `
<Fragment>
  <h3
    className="animate__animated animate__fadeIn"
  >
    Hunter x Hunter
  </h3>
  <p
    className="animate__animated animate__flash"
  >
    Loading...
  </p>
  <div
    className="card-grid"
  />
</Fragment>
`;

exports[`GifGrid component tests should show items when the images are loaded - useFetchGifs 1`] = `
<Fragment>
  <h3
    className="animate__animated animate__fadeIn"
  >
    Hunter x Hunter
  </h3>
  <div
    className="card-grid"
  >
    <GifGridItem
      id="123"
      key="123"
      title="A title"
      url="https://localhost/something.jpg"
    />
  </div>
</Fragment>
`;
